### PR TITLE
Expose block filters over REST

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This adds a new rest endpoint:
/rest/blockfilter/filtertype/requesttype/blockhash (eg
/rest/blockfilter/basic/header/000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f.hex)
which exposes either the filter "header" or the filter data itself.
Most of the code is cribbed from the equivalent RPC.

You can test it at http://bitcoin-rest.bitcoin.ninja/rest//blockfilter/basic/header/000000005b7a58a939b2636f61fa4ddd62258c5fed57667a35d23f2334c4f86d.hex

### CODE REVIEW

**Comment by promag** [comment]
A test would be nice.

**Comment by practicalswift** [comment]
After reading `doc/REST-interface.md` I'm not entirely clear about the assumed trust boundaries.

What recommendations do we give to our users regarding exposing the REST endpoints publicly? Do the recommendations differ from our recommendations with regards to exposing the JSON-RPC endpoints publicly?

As I've understood it we regard the JSON-RPC interface as as an internal control plane only to be accessible by trusted clients. The assumption we're making from a trust boundary perspective seems to be that we assume that an untrusted clients will never be able to connect to the port serving the JSON-RPC interface.

**Comment by laanwj** [comment]
Concept ACK.

> After reading doc/REST-interface.md I'm not entirely clear about the assumed trust boundaries.

That's a fair question (FWIW the limit has always been: only public data, no complex queries, do not parse JSON as input), but I'd suggest opening a new issue for it. Please keep this one for review of the code changes.

**Comment by practicalswift** [comment]
@laanwj Without knowing if consumers are trusted or not it is pretty hard to review it from a security perspective :)

**Comment by laanwj** [comment]
The REST interface is a lightweight interface for querying public data. Consumers are trusted but less so than on RPC (as they don't authenticate). I still wouldn't recommend exposing it directly to the internet. But maybe it's OK to open it "publicly" inside some VPN that your applications run in.

This is my last general comment on this, please open a new issue if you want to continue this discussion.

File: src/rest.cpp
Comment (emilengler) [inline src/rest.cpp]: Nit, the use of a single line if statement would be better

Comment (TheBlueMatt) [reply src/rest.cpp]: Would be much too long a line to do that.

**Comment by jnewbery** [comment]
Concept ACK

**Comment by TheBlueMatt** [comment]
Added a basic sanity test, redid the way headers work to make it easy to get many of them just like the /headers/ request.

File: src/rest.cpp
Comment (emilengler) [reply src/rest.cpp]: Oh sorry I meant to ignore the brackets...

File: src/rest.cpp
Comment (paymog) [inline src/rest.cpp]: I think it would be useful if this error message provided the user with the valid range.

File: src/rest.cpp
Comment (paymog) [inline src/rest.cpp]: should `2000` be turned into a constant?

File: src/rest.cpp
Comment (paymog) [inline src/rest.cpp]: If I'm reading the code right it looks like json is a valid output format too along with binary and hex.

File: src/rest.cpp
Comment (paymog) [inline src/rest.cpp]: the code for extracting `filtertype`, `index` and `blockhash` is shared between these two functions. Should this extraction code be pulled out into their own functions?

File: src/rest.cpp
Comment (paymog) [inline src/rest.cpp]: nit: can the error message be changed to `"Block " + uriParts[1] + " not found"`?

**Comment by jonasschnelli** [comment]
Concept ACK

File: src/rest.cpp
Comment (TheBlueMatt) [reply src/rest.cpp]: Lets leave DRY'ing up rest.cpp for a separate commit. I played with it a bit and there isn't an obvious solution here that doesn't end up adding more lines, but the whole of REST probably could get DRY'd up a lot especially in the results-providing section.

File: src/rest.cpp
Comment (TheBlueMatt) [reply src/rest.cpp]: Hmm, its the same limits as for headers (the code was copied from that section above). Not sure where to document this given the lack of equivalent documentation like we have in RPC where we can fail with help text. Maybe we should have some kind of much more informative general help system like we do for RPC, though its obviously out of scope here.

File: src/rest.cpp
Comment (TheBlueMatt) [reply src/rest.cpp]: Ehh, its the net limit, and dunno where to put it? Not really worth it.

File: src/rest.cpp
Comment (TheBlueMatt) [reply src/rest.cpp]: Hmm, good catch. Looks like headers is busted too, I fixed both.

File: src/rest.cpp
Comment (TheBlueMatt) [reply src/rest.cpp]: Hmm, its coped from the block code, so to keep it the same everywhere I'll leave it.

File: src/rest.cpp
Comment (laanwj) [reply src/rest.cpp]: similarly, it'd make sense to define a constant for this, and use it in both places, instead of hardcoding 2000

File: src/rest.cpp
Comment (luke-jr) [inline src/rest.cpp]: Use `%u` here instead of `2000`?

**Review by luke-jr — CHANGES_REQUESTED** [review]

**Comment by TheBlueMatt** [comment]
Rebased.

File: src/rest.cpp
Comment (jnewbery) [inline src/rest.cpp:32]: nit: `constrexpr`

File: src/rest.cpp
Comment (jnewbery) [inline src/rest.cpp:291]: Since this is all new code, can you adhere to style guide? if blocks with braces or on same lines, variable names in camel case, etc.

**Review by jnewbery — COMMENTED** [review]
Couple of nits. Otherwise looks good.

Commit order is slightly confused. It adds code in one commit then changes it in later commits. I'd suggest putting the fixes first, and then adding the new block filter code in the final commit.

**Comment by naumenkogs** [comment]
Code review ACK 16d8d2da598a7d234275454f9e053ed074eebf3f.
Perhaps would be good to address the style suggestion above.

**Review by meshcollider — COMMENTED** [review]
Concept ACK. This needs a rebase, I've added a few examples of code that has since changed inline. It'd be good to address the style/commit ordering comments above when you rebase. Also please add the change to `doc/REST-interface.md` that Wladimir suggested above.

**Comment by TheBlueMatt** [comment]
Ah! I'd missed the request for doc updates, sorry about that.

**Comment by TheBlueMatt** [comment]
> concept NACK: i'm mostly of the opinion we should deprecate the rest API and make it the responsibility of a proxy server.

That's a *huge* pain. The REST interface is nice precisely *because* you can build a proxy server for it easy - its like 10 lines of nginx.conf. JSON-RPC isn't a nice protocol that is broadly supported, its basically a Bitcoin Core-only thing at this point.

**Comment by JeremyRubin** [comment]
exposing this rest endpoint over NGINX is precisely how our rest endpoint should not be use, so that only serves to add confidence to my notion that it shouldn't be a part of core. A "proper" proxy to the rest interface would sanitize all requests.

**Review by meshcollider — COMMENTED** [review]
utACK c248fc938d04f2c4418eba907f041099416714f7

**Review by maflcko — APPROVED** [review]
ACK

**Comment by sipa** [comment]
Concept ACK

**Comment by TheBlueMatt** [comment]
Addressed feedback, diff since the previous round of acks:

```
$ git diff-tree -U1 7cf4e22080 2b64fa3251
diff --git a/src/rest.cpp b/src/rest.cpp
index dd1be52f86..18ae350630 100644
--- a/src/rest.cpp
+++ b/src/rest.cpp
@@ -343,3 +343,4 @@ static bool rest_block_notxdetails(const std::any& context, HTTPRequest* req, co
 
-static bool rest_filter_header(const std::any& context, HTTPRequest* req, const std::string& strURIPart) {
+static bool rest_filter_header(const std::any& context, HTTPRequest* req, const std::string& strURIPart)
+{
     if (!CheckWarmup(req))
@@ -414,3 +415,3 @@ static bool rest_filter_header(const std::any& context, HTTPRequest* req, const
     case RetFormat::BINARY: {
-        CDataStream ssHeader(SER_NETWORK, PROTOCOL_VERSION);
+        CDataStream ssHeader(SER_NETWORK, PROTOCOL_VERSION | RPCSerializationFlags());
         for (const uint256& header : filter_headers) {
@@ -424,5 +425,4 @@ static bool rest_filter_header(const std::any& context, HTTPRequest* req, const
     }
-
     case RetFormat::HEX: {
-        CDataStream ssHeader(SER_NETWORK, PROTOCOL_VERSION);
+        CDataStream ssHeader(SER_NETWORK, PROTOCOL_VERSION | RPCSerializationFlags());
         for (const uint256& header : filter_headers) {
@@ -451,6 +451,6 @@ static bool rest_filter_header(const std::any& context, HTTPRequest* req, const
     }
-
 }
 
-static bool rest_block_filter(const std::any& context, HTTPRequest* req, const std::string& strURIPart) {
+static bool rest_block_filter(const std::any& context, HTTPRequest* req, const std::string& strURIPart)
+{
     if (!CheckWarmup(req))
@@ -523,3 +523,2 @@ static bool rest_block_filter(const std::any& context, HTTPRequest* req, const s
     }
-
     case RetFormat::HEX: {
@@ -533,3 +532,2 @@ static bool rest_block_filter(const std::any& context, HTTPRequest* req, const s
     }
-
     case RetFormat::JSON: {
@@ -542,3 +540,2 @@ static bool rest_block_filter(const std::any& context, HTTPRequest* req, const s
     }
-
     default: {
$
```

**Review by jnewbery — COMMENTED** [review]
Nearly ACK. A few minor comments inline. The commit log for the first commit is outdated:

```
    Expose block filters over REST.
    
    This adds a new rest endpoint:
    /rest/blockfilter/filtertype/requesttype/blockhash (eg
    /rest/blockfilter/basic/header/000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f.hex)
    which exposes either the filter "header" or the filter data itself.
    Most of the code is cribbed from the equivalent RPC.
```

The endpoints are now `/rest/blockfilterheaders/<FILTERTYPE>/<COUNT>/<BLOCK-HASH>` and `/rest/blockfilter/<FILTERTYPE>/<BLOCK-HASH>`

**Comment by amovfx** [comment]
I read this in the pr-review-club. I am still learning.

**Comment by MarcoFalke** [comment]
re-ack

This feature was roughly ready when the pull was opened more than two years ago. It's been held back on mostly non-behaviour changing "nits", which still haven't been addressed yet. I don't think they'll be addressed any time soon, so it might be better to fix them in a follow-up at this point. Going to merge to move this forward.

**Comment by jnewbery** [comment]
> It's been held back on mostly non-behaviour changing "nits", which still haven't been addressed yet. I don't think they'll be addressed any time soon, so it might be better to fix them in a follow-up at this point.

@MarcoFalke I'm sure this wasn't your intention, but it seems very inconsistent that for most contributors PRs will be closed if review comments haven't been addressed, but in this case the PR was merged despite review comments not being addressed for many months. It shouldn't be the case that certain contributors can get their PRs merged even if they don't follow the project style guide or respond to review comments.

I also disagree that https://github.com/bitcoin/bitcoin/pull/17631#discussion_r733031180 is a "nit". It's feedback on the interface design and changing it later would be an API break.

**Comment by MarcoFalke** [comment]
>  It shouldn't be the case that certain contributors can get their PRs merged even if they don't follow the project style guide or respond to review comments.

Style-only comments are not mandatory to address. This has always been the case, unless the style issue is so "severe" that a linter enforces it. If there are any style issues here that I missed and should be blocking, someone should add a linter to enforce them. Though, I couldn't find any. 

> I also disagree that #17631 (comment) is a "nit". It's feedback on the interface design and changing it later would be an API break.

Apologies, my bad. I wasn't aware there is a suggestion that requests to change the behaviour.

**Comment by jnewbery** [comment]
> Style-only comments are not mandatory to address. This has always been the case, unless the style issue is so "severe" that a linter enforces it. If there are any style issues here that I missed and should be blocking, someone should add a linter to enforce them. Though, I couldn't find any.

I disagree - all review comments should be addressed, even if that only means responding with a comment "I don't intend to change this because ...". If we don't have a culture where all review comments are responded to, then it falls to the maintainers to make the decision about whether the ignored review comments are important or not, and sometimes important review comments fall through the gaps.

**Comment by MarcoFalke** [comment]
Ok, again apologies for missing the comment(s). I've marked this "Up for grabs", so that someone can address the feedback before the next major release (23.0). If no one does, I'll probably revert the changes.

File: doc/REST-interface.md
Comment (michaelfolkson) [reply doc/REST-interface.md:56]: @TheBlueMatt: Any thoughts on the above?

Comment (dergoegge) [reply doc/REST-interface.md:56]: I think the reason for this PR to use `GET /rest/blockfilterheaders/<FILTERTYPE>/<COUNT>/<BLOCK-HASH><bin|hex|json>`  was that the [**blockheader** endpoint](https://github.com/bitcoin/bitcoin/blob/master/doc/REST-interface.md#blockheaders) already uses the same format. (`GET /rest/headers/<COUNT>/<BLOCK-HASH>.<bin|hex|json>`)

@stickies-v makes a good point but in my opinion staying consistent with the current API was the better move.

**Comment by JeremyRubin** [comment]
It also feels premature given my nack + alternative path to implementing this in https://github.com/bitcoin/bitcoin/pull/23309 not having been reviewed -- were it just my NACK and no alternate, I'd agree that perhpas it was sufficiently addressed, but it was my NACK + Alternative + Acks on that approach from > 1 contributor.

File: doc/REST-interface.md
Comment (MarcoFalke) [reply doc/REST-interface.md:56]: Thanks for taking a look @dergoegge . I was thinking about this and assumed that this was done for consistency as well.

**Comment by MarcoFalke** [comment]
I didn't see any NACK.

Screenshot 2021-12-20 at 17-50-57 Expose block filters over REST by TheBlueMatt · Pull Request #17631 · bitcoin bitcoin


I have no idea how to deal with this, but maybe it could make sense to re-NACK a pull request whenever there is a change (push)? Or maybe use the "GitHub review" and click the red cross, which might not be hidden?

**Comment by MarcoFalke** [comment]
And to comment on the content on the NACK: I don't see how this and #23309 are mutually exclusive.

**Comment by luke-jr** [comment]
I agree with Marco that this has been basically ready for years, even if it might be imperfect. It's not too late for further improvements in followups, but blocking this on it was letting the perfect be the enemy of the good IMO.

That being said, it does seem to be a change in merge policy (IMO for the better), and should probably be discussed with the project and applied consistently.

**Comment by JeremyRubin** [comment]
@MarcoFalke well it's a Concept Nack, so it doesn't get stale with changes (as opposed to an Approach Nack + Concept Ack).

W.r.t mutually exclusive, the contention that  I was holding (and @practicalswift too I think?) is that a REST API exposed for unauthenticated users without a strong statement of the security model is probably something we shouldn't be maintaining or increasing dependency on for core, and a preference for doing it as a layer on top of a more strongly audited and maintained interface (JSON RPC) being superior. So extending the REST API and increasing dependency on it for consumers of Core is counter to the goal of removing it and replacing with 'userland' proxies.

**Comment by sipa** [comment]
@JeremyRubin I think the discussion whether to provide a wrapper, and then later whether or not to deprecate/remove the existing REST interface is mostly independent from the question of what features should be exposed by the REST interface (whether that's provided by bitcoind directly or by a wrapper). Whatever risks exists by having a REST interface in the first place aren't worsened by adding another method.

**Comment by MarcoFalke** [comment]
> I do agree the merging procedure followed here was dubious, though.

Ok, maybe I am being blind, but can someone explain to me what was wrong about the merging procedure?

**Did this have too little review?**

I'd say no. There were:

* Concept ACK laanwj https://github.com/bitcoin/bitcoin/pull/17631#issuecomment-559723792
* Concept ACK jonasschnelli  https://github.com/bitcoin/bitcoin/pull/17631#issuecomment-560159029
* Concept ACK sipa https://github.com/bitcoin/bitcoin/pull/17631#issuecomment-945039954
* Nearly ACK jnewbery https://github.com/bitcoin/bitcoin/pull/17631#pullrequestreview-784707190
* Review ACK naumenkogs https://github.com/bitcoin/bitcoin/pull/17631#issuecomment-630288035
* Review ACK meshcollider https://github.com/bitcoin/bitcoin/pull/17631#pullrequestreview-778070991
* Review ACK myself https://github.com/bitcoin/bitcoin/pull/17631#issuecomment-997682927
* Review ACK dergoegge https://github.com/bitcoin/bitcoin/pull/17631#issuecomment-945175070


If it is not possible to merge a pull request with 8 ACKs, then I honestly don't know how much review is needed to get anything merged. I expect that 99.8% of all pull request have less than 8 ACKs.

**Was this merged before a comment should have been replied to?**

Yes, I missed the comment/question about the interface, and it should have been replied to before merge. Luckily it has been replied to after merge (and in my view resolved). The only unaddressed stuff now is refactoring nit comments.

I don't think the merging procedure changed here. I still think that comments should be replied to before merge. However, I also think that style comments that haven't been replied to should not block a pull request indefinitely. If the overall direction of a pull request has received overall support (which this pull request did), then I think style-nits can be dropped or fixed in a follow-up. If the overall direction of a pull request is not agreed upon, or if the style-nits are rendering a follow-up to be a re-implementation, then I think it should not be merged (not applicable to this pull request).

**Something else?**

Again, I am not seeing anything else wrong, so please elaborate.

**Comment by jnewbery** [comment]
>>> can someone explain to me what was wrong about the merging procedure?

>> Was this merged before a comment should have been replied to?

> Yes

@MarcoFalke - this is all I'm claiming - that review comments should be acknowledged and responded to before merge. The author doesn't have to agree with them or implement them, but if someone has taken the time to review the PR, the author should respond to that review before the PR is merged. That happens in 99% of PRs, and I don't see why this one should have been different.

For what it's worth, I agree with @stickies-v that we should make the REST API RESTful. I don't feel very strongly about it because I don't use the REST API myself, but I think the discussion should be had before merge.

And to be clear, I think you do a fantastic job as a maintainer. In this instance you missed something, but it should never have been your responsibility to assess whether that comment **should** have been ignored. It's the author's responsibility to address all feedback on their PR.

File: doc/REST-interface.md
Comment (stickies-v) [reply doc/REST-interface.md:56]: I agree that consistency is important, and in that light am happy to consider this comment as resolved on this PR. However, I think having just 1 prior case should not prevent us from changing course slightly and adopting better RESTful practices. I think this can nicely be done in a separate PR, so I'm happy to look into that and see if there's an elegant way to remain backwards compatible and still adopt/encourage best practices going forward, specifically on these `/headers/` and `/blockfilterheaders/` endpoints.

**Comment by stickies-v** [comment]
FWIW - I've [marked](https://github.com/bitcoin/bitcoin/pull/17631#discussion_r773150826) my RESTful comment as resolved, since @dergoegge [pointed out](https://github.com/bitcoin/bitcoin/pull/17631#discussion_r772514100) this is consistent with another endpoint. As such, a follow-up PR is probably the more elegant way to go. I'm also happy to incorporate the other unaddressed style change comments on #17631 into this new PR.

I appreciate everyone's efforts in both doing the merging and keeping the process honest. I agree that requiring an author's explicit acknowledgement (or dismissal) of all (non-spam) comments would, in my view, be preferable.

**Comment by MarcoFalke** [comment]
Looks like there is a follow up in #23836, so I've removed the "up for grabs" label again.

**Comment by stickies-v** [comment]
#24098 is now ready for review to address the last unaddressed comments (by myself) on this PR.

**Pull request merged**
